backend/src/apiserver/server/job_server.go

```diff
@@ -81,394 +80,322 @@ type JobServer struct {
 	options         *JobServerOptions
 }
 
-func (s *JobServer) CreateJob(ctx context.Context, request *apiv1beta1.CreateJobRequest) (*apiv1beta1.Job, error) {
-	if s.options.CollectMetrics {
-		createJobRequests.Inc()
+func NewJobServer(resourceManager *resource.ResourceManager, options *JobServerOptions) *JobServer {
+	return &JobServer{resourceManager: resourceManager, options: options}
+}
+
+func (s *JobServer) canAccessJob(ctx context.Context, jobID string, resourceAttributes *authorizationv1.ResourceAttributes) error {
+	if !common.IsMultiUserMode() {
+		// Skip authorization if not multi-user mode.
+		return nil
+	}
+
+	if len(jobID) > 0 {
+		job, err := s.resourceManager.GetJob(jobID)
+		if err != nil {
+			return util.Wrap(err, "Failed to authorize with the job ID")
+		}
+		if len(resourceAttributes.Namespace) == 0 {
+			if len(job.Namespace) == 0 {
+				return util.NewInternalServerError(
+					errors.New("Empty namespace"),
+					"The job doesn't have a valid namespace",
+				)
+			}
+			resourceAttributes.Namespace = job.Namespace
+		}
+		if len(resourceAttributes.Name) == 0 {
+			resourceAttributes.Name = job.DisplayName
+		}
 	}
 
-	err := s.validateCreateJobRequest(request)
+	resourceAttributes.Group = common.RbacPipelinesGroup
+	resourceAttributes.Version = common.RbacPipelinesVersion
+	resourceAttributes.Resource = common.RbacResourceTypeJobs
+
+	err := s.resourceManager.IsAuthorized(ctx, resourceAttributes)
 	if err != nil {
-		return nil, util.Wrap(err, "Validate create job request failed.")
+		return util.Wrap(err, "Failed to authorize with API")
 	}
+	return nil
+}
 
-	if common.IsMultiUserMode() {
-		experimentID := common.GetExperimentIDFromAPIResourceReferences(request.Job.ResourceReferences)
-		if experimentID == "" {
-			return nil, util.NewInvalidInputError("Job has no experiment.")
-		}
-		namespace, err := s.resourceManager.GetNamespaceFromExperimentID(experimentID)
+func (s *JobServer) createJob(ctx context.Context, job *model.Job) (*model.Job, error) {
```

Comment:
I'm hesitant on having v1 and v2 api share the underlying method. For two reasons: 1) this will complicate the deprecation work; 2) if we were going to change v2, v1 has to change as well, and vice versa.

Reply:
The common methods use types from the `model` package, which is common for both API versions. Using a single type (1) ensures consistency functional across versions, (2) decreases engineering costs both short-term and long-term. The (2) point includes easier deprecation, development, and maintenance. Having separate functions for each API versions leads to a very bad pattern and many consequences for users and developers.